numpy/_core/src/multiarray/methods.c

```diff
@@ -2166,32 +2140,13 @@ array_setstate(PyArrayObject *self, PyObject *args)
             return NULL;
         }
     }
-
-    if ((PyArray_FLAGS(self) & NPY_ARRAY_OWNDATA)) {
-        /*
-         * Allocation will never be 0, see comment in ctors.c
-         * line 820
-         */
-        PyObject *handler = PyArray_HANDLER(self);
-        if (handler == NULL) {
-            /* This can happen if someone arbitrarily sets NPY_ARRAY_OWNDATA */
-            PyErr_SetString(PyExc_RuntimeError,
-                            "no memory handler found but OWNDATA flag set");
-            return NULL;
-        }
-        PyDataMem_UserFREE(PyArray_DATA(self), n_tofree, handler);
-        PyArray_CLEARFLAGS(self, NPY_ARRAY_OWNDATA);
-    }
-    Py_XDECREF(PyArray_BASE(self));
-    fa->base = NULL;
-
-    PyArray_CLEARFLAGS(self, NPY_ARRAY_WRITEBACKIFCOPY);
-
-    if (PyArray_DIMS(self) != NULL) {
-        npy_free_cache_dim_array(self);
-        fa->dimensions = NULL;
+    /*
+     * Get rid of everything on self, and then populate with pickle data.
+     */
+    if (dealloc_all_but_self(self) < 0) {
+        return NULL;
```

Comment:
OK, we assume that the only place where `__setitem__` is called is unpickling, which means that the _only_ thing that can happen to `self` on error return is that it is deleted (at this point, the array is completely broken!).

Of course, this can't actually fail in that context, because all of the paths that _might_ fail are actually unreachable.